src/mbgl/renderer/tile_pyramid.cpp

```diff
@@ -322,42 +300,45 @@ std::unordered_map<std::string, std::vector<Feature>> TilePyramid::queryRendered
     }
 
     mapbox::geometry::box<double> box = mapbox::geometry::envelope(queryGeometry);
-    // TODO: Find out why we need a special sorting algorithm here.
-    std::vector<std::reference_wrapper<const RenderTile>> sortedTiles{ renderTiles.begin(),
-                                                                       renderTiles.end() };
-    std::sort(sortedTiles.begin(), sortedTiles.end(), [](const RenderTile& a, const RenderTile& b) {
-        return std::tie(a.id.canonical.z, a.id.canonical.y, a.id.wrap, a.id.canonical.x) <
-            std::tie(b.id.canonical.z, b.id.canonical.y, b.id.wrap, b.id.canonical.x);
-    });
+
+    auto cmp = [](const UnwrappedTileID& a, const UnwrappedTileID& b) {
+        return std::tie(a.canonical.z, a.canonical.y, a.wrap, a.canonical.x) <
+            std::tie(b.canonical.z, b.canonical.y, b.wrap, b.canonical.x);
+    };
+
+    std::map<UnwrappedTileID, std::reference_wrapper<Tile>, decltype(cmp)> sortedTiles{renderTiles.begin(), renderTiles.end(), cmp};
 
     auto maxPitchScaleFactor = transformState.maxPitchScaleFactor();
 
-    for (const RenderTile& renderTile : sortedTiles) {
-        const float scale = std::pow(2, transformState.getZoom() - renderTile.id.canonical.z);
-        auto queryPadding = maxPitchScaleFactor * renderTile.tile.getQueryPadding(layers) * util::EXTENT / util::tileSize / scale;
+    for (const auto& entry : sortedTiles) {
+        const UnwrappedTileID& id = entry.first;
+        Tile& tile = entry.second;
+
+        const float scale = std::pow(2, transformState.getZoom() - id.canonical.z);
```

Comment:
this could be evaluated differently - it might be faster (maybe compiler does it)...
```
scale = 2 ** (getZoom() - id.canonical.z)
      = 2 ** (log2(getScale()) - id.canonical.z)
      = 2**log2(getScale()) / 2 ** id.canonical.z
      = getScale() / (1 << id.canonical.z)
```

Reply:
Great observation, thanks! Included with the latest patch.